Histograms and plotting

nsmith- opened the pull request:
This should in principle obviate histbook.  Consider the `Hist` definition and filling implementation somewhat stable.

nsmith- committed:
Hist tools

Basically a self-contained histbook, but with a narrowed scope
To aid in getting things to matplotlib land as quickly and easily as
possible...

Next step: gallery of plot examples

nsmith- committed:
More proper addition, include scaling and rebin for sparse

Towards coalescing datasets into plot groups, and scaling by cross
sections..
Need tests!!

nsmith- committed:
Make it work right

nsmith- committed:
Add some tests

nsmith- committed:
Merge remote-tracking branch 'origin/master' into hist

nsmith- committed:
I had a lot of bugs

nsmith- committed:
More tests

nsmith- committed:
Hist tests finished

nsmith- committed:
First plotting

nsmith- committed:
Correct error bars

nsmith- committed:
Merge branch 'master' of github.com:CoffeaTeam/fnal-column-analysis-tools into topic-histogram

nsmith- committed:
Fix tests, move matplotlib inside function

importing matplotlib can turn ugly apparently
not sure how to test plotting functionality anyway

nsmith- committed:
Python3 test now passes

lgray commented on tests/test_hist_tools.py:
```diff
@@ -0,0 +1,84 @@
+    h_mascots = h_mascots_1 + h_mascots_2
+    assert h_mascots.project("vocalization", values="h*").project("height").project("mass").project("animal").values()[()] == 1040.
+
+    subphylums = {
```
The plural of Phylum is Phyla

nsmith- committed:
Fix improper biology terminology, and change rebin interface

nsmith- committed:
More hist bugfixes

nsmith- committed:
Single axis array

nsmith- committed:
Slices!

nsmith- committed:
Compare key sets for now

nsmith- committed:
Basic plot options

nsmith- committed:
Make hist.plot an explicit import

So that the rest of hist can be used without matplotlib dependency

nsmith- committed:
Finally convert to generalized dense and sparse axes

nsmith- committed:
Clarify sparse bins

nsmith- committed:
Multidimensional sum

nsmith- committed:
Replace rebin_sparse with group

nsmith- committed:
Support 1D plotting with extra dimension dense or sparse

Sadly this uncovered a bug in the Hist structure, must reduce dense axes
when taking a slice

nsmith- committed:
Demo notebooks

No root files for baconbits-hist because they are large
Will trade out with better demo files in the future

nsmith- committed:
Implement overflow preservation when slicing on dense axes

Also, guard against numeric error in uniform bin slicing

nsmith- committed:
Uniform overflow behavior conventions, make them all work

nsmith- committed:
Some documentation

nsmith- committed:
Export histograms to root files
